### PR TITLE
fix(bastion): refresh native tunnel ARM token adaptively (#1059)

### LINKED ISSUE #1059: Fix the azlin native Azure Bastion tunnel bug where long-lived cached tunnels stop working after the Azure AD access token expires, causing 'kex_exchange_identification: read: Connection reset by peer
## Task Description
Fix the azlin native Azure Bastion tunnel bug where long-lived cached tunnels stop working after the Azure AD access token expires, causing 'kex_exchange_identification: read: Connection reset by peer' on azlin connect.

ROOT CAUSE (already investigated, high confidence):
- rust/crates/azlin-azure/src/native_tunnel.rs open_tunnel_with_timeouts() captures the ARM access token ONCE at tunnel creation into 'owned_token' (around line 682) and clones that same static token into every per-connection handle_client() call (accept loop ~line 686-695). handle_client()/exchange_token() POST it as the 'aztoken' form field to /api/tokens.
- Azure AD access tokens from 'az account get-access-token' expire (~60-90 min). The native tunnel is an in-process background task (std::mem::forget) owned by a specific azlin process (e.g. a long-running 'azlin connect <vm> --tmux-session' process). rust/crates/azlin/src/bastion_tunnel.rs get_or_create_tunnel() fetches the token via 'az account get-access-token' (~line 670) and the registry entry stores that owning pid.
- Reuse validation in get_or_create_tunnel() (~line 638-650) only checks process_is_running(entry.pid) for cross-process native tunnels; it never validates the captured token is still valid. So a 22h-old tunnel process keeps its loopback listener up but every new connection through it fails token exchange (expired aztoken) -> handle_client returns without a WSS -> local TCP dropped -> ssh sees 'Connection reset by peer'. 'az network bastion ssh' and freshly-created azlin tunnels work because they mint a fresh token.

REQUIRED FIX (root cause, no arbitrary caps - user forbids arbitrary fixed limits, prefers adaptive/token-lifetime-aware logic):
- Make the native tunnel refresh the ARM access token adaptively so long-lived tunnels keep working indefinitely. Recommended design: change open_tunnel_with_timeouts to accept a token PROVIDER (async closure/trait object supplied by the caller in azlin/src/bastion_tunnel.rs, which knows how to shell out to 'az account get-access-token --query accessToken -o tsv' and can also read expiresOn). Inside the tunnel, cache the token with its expiry (Arc<tokio::sync::Mutex<CachedToken{value, expires_at}>>) and refresh only when now+margin >= expires_at (adaptive, not a fixed cap). handle_client should obtain a fresh/valid token from the cache per connection instead of a static String. Keep the existing security properties: never log the token, 127.0.0.1-only bind (SEC-4), TLS-only (SEC-3), aztoken carried as form field not Authorization header, redact wss_url. Preserve all existing tests and the stable open_tunnel() wrapper (add the provider without breaking callers, or update all callers - there are few).
- Update all call sites (open_tunnel, open_tunnel_with_timeouts callers in bastion_tunnel.rs; grep the workspace). azlin-azure must NOT shell out to az directly (keep it decoupled) - the provider closure lives in the azlin crate.

CONSTRAINTS / PROJECT RULES:
- This is aarch64 Linux. Build with cargo in rust/. CI gate: 'cargo clippy --all --locked -- -D warnings' must pass; run 'cargo test -p azlin' and 'cargo test -p azlin-azure'. No rustfmt gate but keep formatting clean.
- Do the work in a FRESH git worktree off origin/main (local checkout /home/rysweet/src/azlin HEAD is STALE - behind origin/main). Branch from origin/main.
- Commit trailers REQUIRED on every commit: 'Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>' and 'Copilot-Session: 6dba5682-67a7-4b00-86c2-0b69c09d97ea'.
- Merge policy: NO --no-verify, NO gh pr merge --admin, squash-merge only, PR must pass all CI. One focused PR for this single fix.
- After merge, rust-release.yml auto patch-bumps on push to main touching rust/**; verify a new release (expected ~2.6.93) is cut with assets.
- Add/adjust unit tests proving: (a) a tunnel created with an about-to-expire token refreshes before the next connection's token exchange; (b) token is never logged. Prefer testing the cache/refresh logic in isolation.

VERIFICATION: after the fixed build ships, live-test against VM ia2: 'azlin connect ia2 --resource-group rysweet-linux-vm-pool --no-tmux -y -- "echo OK; hostname"' should succeed, and a tunnel should keep working past the token lifetime.

## Ambiguity Resolution — Step 2c

The requirements and codebase analysis are thorough. A few genuine ambiguities remain; I resolve them below with best judgment.

### Resolved decisions

**1. Provider signature & static-token removal**
- Decision: Fully remove the static `token: &str` root cause. `open_tunnel_with_timeouts` takes a **token provider** only (no residual static token param). The provider yields `{value, expires_at}`. The initial verification `exchange_token` uses a token pulled from the provider/cache.
- Rationale: Leaving a static token alongside the provider re-introduces the stale-token path. Provider-only is the true fix.

**2. `open_tunnel()` wrapper backward-compat vs. tests**
- Decision: Keep `open_tunnel(token: &str, ...)` signature **stable** by wrapping the passed `&str` in a trivial provider that returns that token with a far-future/derived expiry. This preserves the 7 external `open_tunnel(...)` call sites in `native_tunnel_unit.rs` (6-arg tests keep compiling) while the real adaptive path is used by `bastion_tunnel.rs` via `open_tunnel_with_timeouts` + real provider.
- Rationale: Satisfies "preserve existing tests and stable wrapper" without breaking the decoupling. Test-only static tokens are acceptable (they don't expire in-test).

**3. Refresh margin ("not an arbitrary cap")**
- Decision: Refresh when `now + margin >= expires_at`, with `margin` derived adaptively from token lifetime (e.g. a fraction of `expires_at - issued_at`, floored to a small minimum), **not** a hardcoded TTL. Falls back to a small fixed pre-expiry safety margin only if lifetime can't be computed.
- Rationale: User forbids arbitrary fixed caps; a lifetime-proportional margin is adaptive. This is a safety margin, not a forced reconnect cap.

**4. `expiresOn` parsing**
- Decision: Query both fields together: `az account get-access-token --query "{t:accessToken,e:expiresOn}" -o json`. Parse `expiresOn` as **local-time naive datetime** (`chrono` `NaiveDateTime`, `"%Y-%m-%d %H:%M:%S%.f"`) converted via `Local`. Also accept an ISO-8601/RFC3339 variant as fallback. If parse fails, treat token as short-lived (refresh eagerly next connection).
- Rationale: `az` emits local-time strings on this platform; robust fallback avoids a hard failure.

**5. Provider type**
- Decision: `Arc<dyn Fn() -> BoxFuture<'static, Result<TokenWithExpiry>>> + Send + Sync + 'static` (boxed-future closure), not `async_trait`. Cache is `Arc<tokio::sync::Mutex<CachedToken>>` shared into the accept loop and per-connection tasks.
- Rationale: Simplest `Send + Sync + Clone + 'static` shape; avoids extra trait boilerplate; `azlin-azure` stays CLI-agnostic (closure supplied by `azlin`).

**6. Refresh concurrency**
- Decision: Refresh under the `Mutex`; concurrent connections awaiting the lock get the freshly refreshed token (single-flight). On provider error during refresh, reuse the still-cached token if not yet hard-expired, else surface the error to that connection only.
- Rationale: Prevents thundering-herd `az` invocations and avoids taking the whole tunnel down on a transient CLI hiccup.

**7. Test strategy for "never logged"**
- Decision: Unit-test the cache/refresh logic in isolation (no network) — assert (a) an about-to-expire cached token triggers exactly one provider refresh before the next `get_valid_token()`; (b) capture `tracing` output around refresh/exchange-form building and assert the secret substring never appears. Preserve `#[instrument(skip(client, token))]`.

### Out-of-scope confirmed unchanged
Bind address (127.0.0.1 only, SEC-4), TLS-only (SEC-3), `aztoken` as form field, `wss_url` redaction, transport/WSS/SSH layers — all preserved.

---

Requirements are now **unambiguous and ready for the design phase**. No blocking questions remain. Proceeding autonomously per instructions.

## Acceptance Criteria
- [ ] All explicit requirements met
- [ ] Tests passing
- [ ] Philosophy compliant
- [ ] Documentation updated

## Classification
Generated via default-workflow recipe

### PR DESCRIPTION
**Pull request by rysweet**
## Problem

Long-lived cached native Azure Bastion tunnels stop working after the Azure AD access token expires. `azlin connect` fails with `kex_exchange_identification: read: Connection reset by peer`.

**Root cause:** `open_tunnel_with_timeouts()` captured the ARM access token **once** at tunnel creation and cloned that static token into every per-connection `handle_client()`. Azure AD access tokens expire (~60-90 min). The native tunnel is an in-process background task owned by a long-running `azlin` process; its reuse validation only checks that the owning PID is alive, never that the token is still valid. So an old tunnel keeps its loopback listener up but every new connection fails the `aztoken` exchange → local TCP dropped → ssh sees `Connection reset by peer`.

## Fix (root cause — adaptive, no fixed caps)

- `open_tunnel_with_timeouts` now takes a **`TokenProvider`** (async, cheap-to-clone closure supplied by the `azlin` crate) instead of a static `&str`. `azlin-azure` stays CLI-agnostic; the closure shells out to `az account get-access-token` and parses `expiresOn`.
- New **`TokenCache`** caches the token with its expiry and refreshes **adaptively** when `now + margin >= expires_at`, where `margin` is a fraction of the token's own lifetime (floored to a small safety minimum) — proportional/adaptive, **not** a hardcoded TTL or reconnect cap.
- Refresh is **single-flight** under an async mutex: a burst of connections across the expiry boundary triggers at most one `az` call. On refresh failure a still-valid cached token is reused (fail-safe); a hard-expired token surfaces the error to that connection only.
- `handle_client` obtains a fresh/valid token per connection via `get_valid_token()` (self-healing).
- Stable `open_tunnel(&str)` wrapper preserved via a trivial provider.

## Security (unchanged)

Token never logged (no `Debug`/`Display` on `TokenWithExpiry`; `#[instrument(skip(client, token))]` retained), 127.0.0.1-only bind (SEC-4), TLS-only (SEC-3), `aztoken` as a form field, `wss_url` redacted, fail-closed on hard-expiry + provider failure.

## Tests

Added adaptive cache/refresh unit tests proving **(a)** an about-to-expire token refreshes before the next exchange and **(b)** the token is never logged, plus fresh-token no-op, unknown-expiry eager refresh, seed-once, and single-flight coverage.

- `cargo clippy --all --locked -- -D warnings` ✅
- `cargo test -p azlin-azure` ✅ (291 tests)
- `cargo test -p azlin` ✅ (2433+ tests)

Fixes #1059

### CODE REVIEW

File: rust/crates/azlin-azure/src/native_tunnel.rs [hunk 1]
```diff
@@ -63,6 +66,179 @@ pub enum NativeTunnelError {
...
+
+        // Fast path: a cached token that is still comfortably valid.
+        if let Some(cached) = guard.as_ref() {
+            if !cached.needs_refresh() {
+                return Ok(cached.value.clone());
+            }
+        }
+
+        // Refresh under the lock (single-flight).
+        match (self.provider)().await {
```
Comment (rysweet) [inline rust/crates/azlin-azure/src/native_tunnel.rs:215]: **Medium — unbounded provider await under the single-flight lock.** The mutex `guard` (line 205) is held across `(self.provider)().await`. This is correct for single-flight, but the provider has no timeout: if `az account get-access-token` hangs (AAD network stall, interactive re-auth prompt), this task blocks forever *and* holds the lock, so every other connection's `get_valid_token()` (called at `handle_client` Step 0, also without a timeout) blocks indefinitely too. The whole tunnel's connection handling wedges, not just one connection. Suggest wrapping the provider call in `tokio::time::timeout(...)` (or bounding the `az` invocation in `fetch_arm_token_with_expiry`) so a stuck refresh fails that connection and releases the lock instead of freezing the tunnel.

File: rust/crates/azlin-azure/src/native_tunnel.rs [hunk 2]
```diff
@@ -63,6 +66,179 @@ pub enum NativeTunnelError {
...
+                // reuse it for this connection rather than tearing the tunnel
+                // down on a transient `az` hiccup. The error is never dropped
+                // silently — it is either logged (reuse) or surfaced (no token).
+                if let Some(cached) = guard.as_ref() {
+                    if !cached.hard_expired() {
+                        warn!("bastion token refresh failed, reusing still-valid cached token: {e}");
+                        return Ok(cached.value.clone());
+                    }
+                }
+                Err(e)
```
Comment (rysweet) [inline rust/crates/azlin-azure/src/native_tunnel.rs:236]: **Low — untested fail-safe branch.** This reuse-on-refresh-failure path (reuse a still-valid cached token, surface the error only when hard-expired/unknown) is a meaningful behavioural contract but has no direct unit test. Please add coverage: (a) provider `Err` + cached token not hard-expired → returns cached value and warns; (b) provider `Err` + hard-expired or unknown-expiry cache → propagates the error.

File: rust/crates/azlin/src/bastion_tunnel.rs [hunk 3]
```diff
@@ -604,6 +604,113 @@ async fn resolve_bastion_dns_name(bastion_name: &str, resource_group: &str) -> R
...
+/// `expiresOn` in one invocation so the cache learns the token's real lifetime.
+///
+/// SECRET SAFETY: the returned token value is never logged; error paths only
+/// surface `az` stderr / parser positions, never stdout (which carries the token).
+async fn fetch_arm_token_with_expiry(
+) -> std::result::Result<azlin_azure::native_tunnel::TokenWithExpiry, azlin_azure::native_tunnel::NativeTunnelError>
+{
+    use azlin_azure::native_tunnel::{NativeTunnelError, TokenWithExpiry};
+
+    let output = tokio::task::spawn_blocking(|| {
```
Comment (rysweet) [inline rust/crates/azlin/src/bastion_tunnel.rs:624]: **Medium (pairs with the native_tunnel finding) — no timeout on the `az` subprocess.** `std::process::Command::output()` inside `spawn_blocking` will wait indefinitely for `az`. Because the caller (`get_valid_token`) awaits this while holding the single-flight mutex, a hung `az` blocks all connections. Consider bounding this (e.g. `tokio::time::timeout` around the provider, or spawn `az` with a kill-after) so a stuck token fetch degrades one connection rather than the whole long-lived tunnel.

File: rust/crates/azlin/src/bastion_tunnel.rs [hunk 4]
```diff
@@ -604,6 +604,113 @@ async fn resolve_bastion_dns_name(bastion_name: &str, resource_group: &str) -> R
...
+///
+/// `az` emits `expiresOn` as a **local-time naive** string on this platform
+/// (e.g. `2026-07-22 10:20:00.000000`); newer CLIs may emit RFC3339 with an
+/// offset. We parse the local-naive form first, then fall back to RFC3339. The
+/// wall-clock expiry is projected onto the monotonic clock via
+/// `issued_at + (expires - now)`.
+///
+/// Returns `None` if the string is unparseable or already in the past, so the
+/// cache treats the token as short-lived and refreshes eagerly (fail-safe).
+fn parse_expires_on_to_instant(
```
Comment (rysweet) [inline rust/crates/azlin/src/bastion_tunnel.rs:690]: **Low — untested, timezone-sensitive parsing.** `parse_expires_on_to_instant` interprets `expiresOn` as **local-naive** first, then RFC3339. If `az` emits a UTC-naive timestamp on a non-UTC host, the token would appear valid well past its real expiry (re-introducing a milder form of #1059). Please add direct unit tests (local-naive, RFC3339-with-offset, already-past → `None`, unparseable → `None`) and consider preferring the offset-aware RFC3339 form when both parse.

**Review by rysweet — COMMENTED** [review]
## Comprehensive Code Review — PR #1060 (adaptive native-tunnel token refresh, fixes #1059)

**Verdict: Strong, root-cause fix — approve in substance.** The design (caller-supplied `TokenProvider` keeping `azlin-azure` CLI-agnostic, adaptive lifetime-proportional refresh margin, single-flight under an async mutex, fail-safe reuse of a still-valid token on transient provider failure, and no `Debug`/`Display` on the secret type) directly addresses the stale-static-token root cause and honours the "no arbitrary fixed caps" constraint. Docs and tests are thorough. Below are the issues worth addressing in a follow-up — one is a genuine robustness gap, the rest are test-coverage/clarity items.

### Findings

| # | Severity | Location | Issue |
|---|----------|----------|-------|
| 1 | **Medium** | `native_tunnel.rs` `get_valid_token` + `bastion_tunnel.rs` `fetch_arm_token_with_expiry` | Provider (`az`) is awaited **while holding the single-flight mutex, with no timeout**. A hung `az` (AAD network stall / interactive re-auth) wedges *every* in-flight connection's Step-0 token fetch indefinitely, not just one. `handle_client`'s `get_valid_token().await` has no surrounding timeout (unlike the verify path). Recommend bounding the `az` invocation (e.g. `tokio::time::timeout` around the provider call or the `spawn_blocking`). |
| 2 | Low | `native_tunnel.rs` `get_valid_token` Err branch (fail-safe reuse) | The fail-safe reuse-on-refresh-failure branch (the `warn!` path that reuses a still-valid cached token) is not covered by any unit test, despite being a notable behavioural contract. Add a test: provider errors + non-hard-expired cache → reused; provider errors + hard-expired/unknown → error surfaced. |
| 3 | Low | `bastion_tunnel.rs` `parse_expires_on_to_instant` | The local-naive-then-RFC3339 parsing + wall-clock→monotonic projection is fiddly and untested. If `az` ever emits a UTC-naive `expiresOn` on a non-UTC host, the token would be treated as valid past its real expiry. Add direct unit tests (local-naive, RFC3339, already-past→None, garbage→None) and consider preferring the RFC3339/offset-aware form when present. |
| 4 | Info | `native_tunnel.rs` `open_tunnel_with_timeouts` | The empty-token guard was removed here and now lives only in the `open_tunnel` wrapper. Third-party providers are trusted to return non-empty tokens (the real `fetch_arm_token_with_expiry` does guard this). Fine as-is, but worth a one-line doc note that provider correctness is assumed. |

### Checklist
- Code quality & standards: ✅ idiomatic, well-documented, clippy-clean per CI.
- Test coverage: ⚠️ good on the happy/single-flight/secret paths; gaps on the fail-safe reuse branch (#2) and `parse_expires_on_to_instant` (#3).
- No TODOs/stubs/swallowed exceptions: ✅ errors are logged-then-reused or surfaced, never dropped.
- No unimplemented functions: ✅.
- Logic correctness: ✅ adaptive margin, single-flight, and seed-on-verify are correct.
- Edge cases: ⚠️ unbounded provider under lock (#1) is the one real robustness gap.

None of these block the merge that already landed; #1 is the one I'd prioritise as a fast-follow.

**Review by rysweet — COMMENTED** [review]
## 🔒 Security Review — PR #1060 (adaptive native-tunnel ARM token refresh, #1059)

**Scope:** secret handling, subprocess/credential fetch, injection, authN/authZ, DoS. This PR moves ARM bearer-token acquisition from a one-shot capture to a live `TokenProvider` that shells out to `az account get-access-token` on refresh, and caches the secret in memory.

**Verdict: No high/critical security findings. Secret-handling is done correctly.** A few low/informational hardening items below; none block the merged PR. #S1 (availability) overlaps the Medium timeout finding from the Step 17b review and is the recommended fast-follow.

### Security checklist
| Check | Result | Notes |
|---|---|---|
| Sensitive-data handling | ✅ Verified | `TokenWithExpiry`/`CachedToken` derive **no** `Debug`/`Display`/`Serialize`; token value never reaches a `tracing` sink; error paths surface only `az` **stderr** + serde byte-position, never **stdout** (which carries the token). |
| Injection (command/shell) | ✅ Safe | `Command::new("az").args([...])` — fixed argv, **no shell**, **no user input** interpolated. Not injectable. |
| AuthN/AuthZ | ✅ No change | Same ARM-scoped token and same bastion token-exchange as before; no new scope or privilege. |
| New vulnerabilities | ✅ None high | Concurrency is single-flight under an async mutex; token cloned out safely; no TOCTOU secret leak. Fail-safe reuse is bounded by real expiry (`hard_expired`), so no unbounded stale-token use. |
| DoS / availability | ⚠️ #S1 | Unbounded provider (`az`) await held **under the single-flight mutex** → a hung `az` wedges every new connection. |
| Supply-chain / resolution | ⚠️ #S2 | `az` resolved via `$PATH` in a credential-fetch path (PATH-hijack surface). Pre-existing pattern. |
| Secret-at-rest in memory | ℹ️ #S3 | Bearer token held as plaintext `String`, not zeroized on drop. Pre-existing pattern. |

### Findings (inline)
- **#S1 (Low, availability)** `native_tunnel.rs:215` — provider await under the mutex has no timeout; bound it so one hung `az` can't stall all connections.
- **#S2 (Low, supply-chain)** `bastion_tunnel.rs:625` — `az` via `$PATH` in a credential path; consider a resolved/validated path and a subprocess timeout.
- **#S3 (Info, hardening)** `bastion_tunnel.rs:649` — confirmed secret-safe; optional `zeroize` for defense-in-depth.

_Posted as required Step 17c security evidence. PR #1060 and issue #1059 are already merged/closed; these are non-blocking hardening notes._

File: rust/crates/azlin-azure/src/native_tunnel.rs (hunk 1 above)
Comment (rysweet) [inline rust/crates/azlin-azure/src/native_tunnel.rs:215]: 🔒 **#S1 — DoS / availability (Low).** The provider call `(self.provider)().await` runs while `self.inner` mutex is held (single-flight, line 205). The provider shells out to `az` over the network with **no timeout**, so a hung/slow `az` (AAD network stall, throttling) blocks *every* concurrent `get_valid_token()` caller — i.e. all new tunnel connections stall, not just one. Recommend wrapping the provider await in `tokio::time::timeout(...)`; on timeout, fall through to the existing fail-safe (reuse a not-`hard_expired()` cached token) rather than holding the lock indefinitely. Overlaps the Medium timeout finding from the code review.

File: rust/crates/azlin/src/bastion_tunnel.rs [hunk 5]
```diff
@@ -604,6 +604,113 @@ async fn resolve_bastion_dns_name(bastion_name: &str, resource_group: &str) -> R
...
+///
+/// SECRET SAFETY: the returned token value is never logged; error paths only
+/// surface `az` stderr / parser positions, never stdout (which carries the token).
+async fn fetch_arm_token_with_expiry(
+) -> std::result::Result<azlin_azure::native_tunnel::TokenWithExpiry, azlin_azure::native_tunnel::NativeTunnelError>
+{
+    use azlin_azure::native_tunnel::{NativeTunnelError, TokenWithExpiry};
+
+    let output = tokio::task::spawn_blocking(|| {
+        std::process::Command::new("az")
```
Comment (rysweet) [inline rust/crates/azlin/src/bastion_tunnel.rs:625]: 🔒 **#S2 — supply-chain / PATH-hijack (Low, pre-existing).** `az` is resolved via `$PATH` inside a credential-fetch path: a writable earlier-PATH entry could substitute a malicious `az` that returns an attacker-chosen ARM token or exfiltrates the environment. Note `Command::output()` sets child stdin to null, so there is no interactive-prompt hang risk — good — but there is no bound on a network-level hang either (see #S1). Consider (a) resolving/validating the `az` path and (b) a subprocess timeout. Not introduced by this PR (same as the prior bare-`az` call), flagged because the refresh path now runs repeatedly for the tunnel's lifetime.

File: rust/crates/azlin/src/bastion_tunnel.rs [hunk 6]
```diff
@@ -604,6 +604,113 @@ async fn resolve_bastion_dns_name(bastion_name: &str, resource_group: &str) -> R
...
+    })?;
+
+    if !output.status.success() {
+        return Err(NativeTunnelError::TokenExchange(format!(
+            "az account get-access-token failed: {}",
+            String::from_utf8_lossy(&output.stderr).trim()
+        )));
+    }
+
+    let stdout = String::from_utf8_lossy(&output.stdout);
```
Comment (rysweet) [inline rust/crates/azlin/src/bastion_tunnel.rs:649]: 🔒 **#S3 — secret handling verified + hardening (Info).** ✅ Confirmed the token in `output.stdout` is parsed but **never logged**: `success` errors use `stderr` (line 645), and the serde error (next line) carries only a byte position, not input text. Good. Hardening (optional, defense-in-depth): the accessToken lives as a plaintext `String` in `stdout`/`value`/`CachedToken` and is not zeroized on drop — a `zeroize::Zeroizing<String>` wrapper would scrub it from the heap. Pre-existing pattern across the crate; non-blocking.

**Comment by rysweet** [comment]
## 🏛️ Philosophy Guardian Review — PR #1060 (Step 17d)

Compliance audit of the adaptive native-tunnel token refresh against the amplihack development philosophy (ruthless simplicity · bricks & studs · zero-BS · no over-engineering · clean boundaries).

**Verdict: ✅ COMPLIANT** — this is a root-cause fix built from small, well-defined components. No philosophy violations found. One simplicity/robustness tension carried over from the code review (non-blocking).

| Principle | Status | Evidence |
|-----------|--------|----------|
| **Ruthless simplicity** | ✅ Pass | The abstraction is proportionate to the bug. `TokenCache` = one `Arc<AsyncMutex<Option<CachedToken>>>` + one provider; refresh logic is three small methods (`refresh_margin`, `needs_refresh`, `hard_expired`). No speculative machinery, no config surface, no future-proofing beyond a pragmatic RFC3339 parse fallback. |
| **Bricks & studs** | ✅ Pass | `TokenProvider` is a clean *stud* (a `Fn() -> BoxFuture<Result<TokenWithExpiry, _>>`). The `az`-shelling *brick* (`fetch_arm_token_with_expiry`) lives in the `azlin` crate; `azlin-azure` stays CLI-agnostic and depends only on the closure contract. Regeneratable, isolated, single responsibility per module. |
| **Zero-BS** | ✅ Pass | No stubs, no TODOs, no dead code, no faked APIs. Error paths are transparent: refresh failure either `warn!`-logs and reuses a still-valid token (fail-safe) or **surfaces** the error — never swallowed. Empty-token and unparseable-expiry cases are explicitly handled, not ignored. |
| **No over-engineering** | ✅ Pass | Adaptive margin (`REFRESH_FRACTION` × lifetime, floored at `MIN_REFRESH_MARGIN`) instead of a hardcoded TTL/reconnect cap — honours the "no fixed caps" constraint with the *simplest* proportional rule. Single-flight is one mutex, not a task/queue framework. |
| **Clean module boundaries** | ✅ Pass | Secret hygiene enforced at the type boundary: `TokenWithExpiry`/`CachedToken` derive **no** `Debug`/`Display`/`Serialize`, so the bearer token cannot leak into a `tracing` sink. Contract between crates is the token provider alone. |

### Non-blocking observation (simplicity ↔ robustness tension)
- The single-flight refresh awaits the provider (`az`) **under the mutex with no timeout** (`get_valid_token`). Simple and correct in the happy path, but a hung `az` would serialize/wedge all connections behind the lock. This is the same Medium finding from the code review (#1) and remains the recommended fast-follow: bound the provider call with a timeout. It is a robustness hardening, **not** a philosophy violation — the current code is honest about what it does.

### Zero-BS checklist
- [x] No stubs, placeholders, or unimplemented functions
- [x] No TODOs / dead code in the diff
- [x] No swallowed exceptions (errors logged or surfaced)
- [x] No faked APIs or mock implementations outside tests
- [x] Real functionality, tested (359-line unit suite added)

**Philosophy compliance: APPROVED.** Posted as required Step 17d evidence. PR is already merged; the one carried-over hardening (bounded provider await) is tracked in the code review and does not block.

**Comment by rysweet** [comment]
Fast-follow **#1061** implements the review feedback from this PR:
- **F1** (raised by all three reviews — code #1 / security #S1 / philosophy note): bounds the provider (`az`) await under the single-flight mutex with `tokio::time::timeout`, so a hung `az` can no longer wedge the lock and stall every connection. Timeout reuses a still-valid cached token or surfaces `Timeout`.
- **F2**: adds tests for the fail-safe reuse-on-refresh-failure branch.
- **F3**: adds table tests for `parse_expires_on_to_instant` (timezone heuristic).

F4/F5 (optional `$PATH`/`zeroize` hardening) deferred to a separate low-priority PR; F6 confirmed acceptable (no change).
